### PR TITLE
Ginkgo: Migrated LB services tests

### PR DESCRIPTION
**Pull request by eloycoto**
Cover the changes made in PR #1674  

Only applies the services ones, the other ones are already covered by others PR.

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

### CODE REVIEW

**Comment by ianvernon** [comment]
Please reference which file / block of code this covers the changes made as part of PR #1674 ? Include line numbers if possible. Thanks!

**Comment by eloycoto** [comment]
Hi @ianvernon 


Few things here, the following changes were not applied, it was fixed in ginkgo during last months: 
- 01-ct --> Disable wait for endpoints
- 08-nat46 --> Fix name
- 11-getting-started --> Disabled wait_for_cilium
- 14-policy-enforcement-docker.sh -> Disabled wait_for_endpoints
- 16-cidr-ingress-policy.sh --> Disabled policy_delete_and_wait

The big changes were in LB(10-proxy), something that was covered by ginkgo Policies direct mode (Pod to Pod) but no using LB. The lines that were moved are 10-proxy.sh:292-325. As I said, no direct-direct testing, it was done in `runtime/Policies.go`. The policies are not equal but have the same behaviour. 

Regards.

File: test/runtime/lb.go [hunk 1]
```diff
@@ -301,6 +301,103 @@ var _ = Describe("RuntimeLB", func() {
...
+				return helpers.CurlFail(fmt.Sprintf(
+					"http://%s/%s", service, target))
+			}
+
+			status := vm.ServiceAdd(100, service1, []string{
+				fmt.Sprintf("%s:80", httpd1["IPv4"])}, 1)
+			status.ExpectSuccess("L4 service cannot be created")
+
+			status = vm.ServiceAdd(101, service2, []string{
+				fmt.Sprintf("%s:80", httpd2["IPv4"])}, 1)
```
Comment (ianvernon) [inline test/runtime/lb.go]: Use string IPv4 from `cons.go` instead of hardcoding it throughout this file.

File: test/runtime/lb.go [hunk 2]
```diff
@@ -301,6 +301,103 @@ var _ = Describe("RuntimeLB", func() {
...
+			status.ExpectSuccess("L4 service cannot be created")
+
+			_, err = vm.PolicyImport(vm.GetFullPath(policiesL7JSON), helpers.HelperTimeout)
+			Expect(err).Should(BeNil())
+
+			By("Simple Ingress")
+			status = vm.ContainerExec(helpers.App1, getHTTP(service1, "public"))
```
Comment (ianvernon) [inline test/runtime/lb.go]: Please make "public" and "private" constants in `cons.go` and utilize them in this test.

File: test/runtime/lb.go [hunk 3]
```diff
@@ -301,6 +301,103 @@ var _ = Describe("RuntimeLB", func() {
 		status = vm.ContainerExec("client", testCmd)
 		status.ExpectSuccess("LB is not working after restart")
 	})
+
+	Context("Services Policies", func() {
+		BeforeEach(func() {
+			vm.PolicyDelAll()
```
Comment (ianvernon) [inline test/runtime/lb.go]: Should all services also be deleted in `BeforeEach`? Would ensure that there's no extra cruft lying around beforehand.

File: test/runtime/lb.go [hunk 4]
```diff
@@ -301,6 +301,103 @@ var _ = Describe("RuntimeLB", func() {
...
+		})
+
+		AfterEach(func() {
+			vm.SampleContainersActions(helpers.Delete, helpers.CiliumDockerNetwork)
+			vm.PolicyDelAll()
```
Comment (ianvernon) [inline test/runtime/lb.go]: Why is the `*CmdRes` from `PolicyDelAll` being ignored here? We should either explicitly ignore the output of functions using `_` and add a comment explaining why we are ignoring the return value(s) of a function or check the return values for specific conditions. Silently ignoring output might make debugging of tests more difficult.

File: test/runtime/lb.go [hunk 5]
```diff
@@ -301,6 +301,103 @@ var _ = Describe("RuntimeLB", func() {
...
+			By("Multiple Egress")
+
+			status = vm.ContainerExec(helpers.App2, getHTTP(service2, "public"))
+			status.ExpectSuccess()
+
+			status = vm.ContainerExec(helpers.App2, getHTTP(service2, "private"))
+			status.ExpectFail()
+		}
+
+		It("Contrack enabled", func() {
```
Comment (ianvernon) [inline test/runtime/lb.go]: Contrack --> Conntrack

File: test/runtime/lb.go [hunk 6]
```diff
@@ -301,6 +301,103 @@ var _ = Describe("RuntimeLB", func() {
...
+			status := vm.ExecCilium(fmt.Sprintf("config %s=true",
+				helpers.OptionConntrackLocal))
+			status.ExpectSuccess()
+
+			vm.SampleContainersActions(helpers.Create, helpers.CiliumDockerNetwork)
+			vm.WaitEndpointsReady()
+			policiesTest()
+		})
+
+		It("Contrack disabled", func() {
```
Comment (ianvernon) [inline test/runtime/lb.go]: Contrack --> Conntrack

File: test/runtime/lb.go [hunk 7]
```diff
@@ -301,6 +301,103 @@ var _ = Describe("RuntimeLB", func() {
...
+			status := vm.ExecCilium(fmt.Sprintf("config %s=false",
+				helpers.OptionConntrackLocal))
+			status.ExpectSuccess()
+
+			vm.SampleContainersActions(helpers.Create, helpers.CiliumDockerNetwork)
+			vm.WaitEndpointsReady()
```
Comment (ianvernon) [inline test/runtime/lb.go]: Why is the returned `bool` from `WaitEndpointsReady()` being ignored here? We should either explicitly ignore the output of functions using `_` and add a comment explaining why we are ignoring the return value(s) of a function or check the return values for specific conditions. Silently ignoring output might make debugging of tests more difficult.

File: test/runtime/lb.go [hunk 8]
```diff
@@ -301,6 +301,103 @@ var _ = Describe("RuntimeLB", func() {
...
+			status.ExpectFail()
+		}
+
+		It("Contrack enabled", func() {
+			status := vm.ExecCilium(fmt.Sprintf("config %s=true",
+				helpers.OptionConntrackLocal))
+			status.ExpectSuccess()
+
+			vm.SampleContainersActions(helpers.Create, helpers.CiliumDockerNetwork)
+			vm.WaitEndpointsReady()
```
Comment (ianvernon) [inline test/runtime/lb.go]: Why is the returned `bool` from `WaitEndpointsReady()` being ignored here? We should either explicitly ignore the output of functions using `_` and add a comment explaining why we are ignoring the return value(s) of a function or check the return values for specific conditions. Silently ignoring output might make debugging of tests more difficult.

**Review by ianvernon — CHANGES_REQUESTED** [review]

**Review by ianvernon — APPROVED** [review]

File: test/runtime/lb.go [hunk 9]
```diff
@@ -301,6 +301,108 @@ var _ = Describe("RuntimeLB", func() {
 		status = vm.ContainerExec("client", testCmd)
 		status.ExpectSuccess("LB is not working after restart")
 	})
+
+	Context("Services Policies", func() {
+		BeforeEach(func() {
+			vm.PolicyDelAll().ExpectSuccess()
+			vm.ServiceDelAll().ExpectSuccess()
```
Comment (jrajahalme) [inline test/runtime/lb.go]: Should not be necessary to delete the services and policies both before and after each test?

File: test/runtime/lb.go [hunk 10]
```diff
@@ -301,6 +301,108 @@ var _ = Describe("RuntimeLB", func() {
...
+		})
+
+		AfterEach(func() {
+			vm.SampleContainersActions(helpers.Delete, helpers.CiliumDockerNetwork)
+			vm.PolicyDelAll().ExpectSuccess()
+			vm.ServiceDelAll().ExpectSuccess()
+
+			status := vm.ExecCilium(fmt.Sprintf("config %s=false",
+				helpers.OptionConntrackLocal))
+			status.ExpectSuccess()
```
Comment (jrajahalme) [inline test/runtime/lb.go:317]: Resetting the conntrack option after each test seems also a bit excessive, or maybe I'm not understanding the Ginkgo semantics correctly?

File: test/runtime/lb.go [hunk 11]
```diff
@@ -301,6 +301,108 @@ var _ = Describe("RuntimeLB", func() {
...
+		policiesTest := func() {
+
+			httpd1, err := vm.ContainerInspectNet(helpers.Httpd1)
+			Expect(err).Should(BeNil())
+
+			httpd2, err := vm.ContainerInspectNet(helpers.Httpd2)
+			Expect(err).Should(BeNil())
+
+			service1 := "2.2.2.100:80"
+			service2 := "2.2.2.101:80"
```
Comment (jrajahalme) [inline test/runtime/lb.go]: To replicate the tests in 10-proxy.sh, one of the service addresses should be IPv4 and the other IPv6. Both should then use the same two back end servers.

File: test/runtime/lb.go [hunk 12]
```diff
@@ -301,6 +301,108 @@ var _ = Describe("RuntimeLB", func() {
...
+			status = vm.ContainerExec(helpers.App1, getHTTP(service1, helpers.Public))
+			status.ExpectSuccess()
+
+			status = vm.ContainerExec(helpers.App2, getHTTP(service1, helpers.Public))
+			status.ExpectFail()
+
+			By("Simple Egress")
+
+			status = vm.ContainerExec(helpers.App2, getHTTP(service2, helpers.Public))
+			status.ExpectSuccess()
```
Comment (jrajahalme) [inline test/runtime/lb.go:357]: Would App1 to service2 fail? Maybe add that?

File: test/runtime/lb.go [hunk 13]
```diff
@@ -301,6 +301,108 @@ var _ = Describe("RuntimeLB", func() {
...
+			status = vm.ContainerExec(helpers.App2, getHTTP(service1, helpers.Public))
+			status.ExpectFail()
+
+			By("Multiple Egress")
+
+			status = vm.ContainerExec(helpers.App2, getHTTP(service2, helpers.Public))
+			status.ExpectSuccess()
+
+			status = vm.ContainerExec(helpers.App2, getHTTP(service2, helpers.Private))
+			status.ExpectFail()
```
Comment (jrajahalme) [inline test/runtime/lb.go:383]: Add App1 -> service2 cases?

File: test/runtime/lb.go [hunk 14]
```diff
@@ -301,6 +301,108 @@ var _ = Describe("RuntimeLB", func() {
...
+			Expect(ready).To(BeTrue())
+			policiesTest()
+		})
+
+		It("Conntrack disabled", func() {
+			status := vm.ExecCilium(fmt.Sprintf("config %s=false",
+				helpers.OptionConntrackLocal))
+			status.ExpectSuccess()
+
+			vm.SampleContainersActions(helpers.Create, helpers.CiliumDockerNetwork)
```
Comment (jrajahalme) [inline test/runtime/lb.go:402]: Why does the Docker network need to be recreated?

File: test/runtime/lb.go (hunk 10 above)
Comment (eloycoto) [reply test/runtime/lb.go:317]: It's a Ginkgo thing, need to happens after each test :-(

File: test/runtime/lb.go (hunk 14 above)
Comment (eloycoto) [reply test/runtime/lb.go:402]: That method does not create a network, it creates the sample containers in the given docker network.

File: test/runtime/lb.go (hunk 12 above)
Comment (eloycoto) [reply test/runtime/lb.go:357]: Is an Egress policy from App2, I've just added app2 to private ;-) App1 will always work.

File: test/runtime/lb.go (hunk 13 above)
Comment (eloycoto) [reply test/runtime/lb.go:383]: On egress will always work and it is already tested on runtime/Policies.go

**Comment by manalibhutiyani** [comment]
If all review comments are addressed, can we merge this?

**Comment by raybejjani** [comment]
@eloycoto Could you rebase this branch? I will then merge it.

**Comment by eloycoto** [comment]
@raybejjani done!

**Pull request merged**